runtime/shared_common/CompositeCache.cpp

```diff
@@ -6813,7 +6813,7 @@ SH_CompositeCacheImpl::updateRuntimeFullFlags(J9VMThread *currentThread)
 	if (_readOnlyOSCache || J9_ARE_ALL_BITS_SET(*_runtimeFlags, J9SHR_RUNTIMEFLAG_ENABLE_READONLY)) {
 		goto done;
 	}
-
+	cacheFullFlags = _theca->cacheFullFlags;
```

Comment:
This isn't a fix, you can use compare and swap to fix it.

Reply:
You mean compare and swap on local variable cacheFullFlags ?

Reply:
The time window I find in the old code is between line 6802 and line 6817. Line 6802 got a value of 0 from  _theca->cacheFullFlags.  When doing comparison at line 6817, both _theca->cacheFullFlags and _cacheFullFlags are changed by another thread. The current thread is still using a cached cacheFullFlags which is 0 .

Reply:
Compare and swap can be used to combine the test `_cacheFullFlags == cacheFullFlags` and the assignment `_cacheFullFlags = cacheFullFlags` into one operation. Sounds like the fetch of `_theca->cacheFullFlags`, the test and the assignment need to be protected by a monitor. Is the code that changed `_theca->cacheFullFlags` in another thread using any monitor?

Reply:
The other thread is changing _theca->cacheFullFlags, _cacheFullFlags  and the runtime flags here:
 https://github.com/eclipse/openj9/blob/master/runtime/shared_common/CompositeCache.cpp#L6142to#L6150
It is not using a monitor. But it is inside the write mutex. 

We shouldn't cache the obsolete _theca->cacheFullFlags value to the local variable cacheFullFlags and use the obsolete value to update the runtime flags.  I think we should fetch the latest _theca->cacheFullFlags to update the runtime flags.

Reply:
Since a thread switch can occur at any time, and the cpu is free to move code around somewhat, changing the source to move the fetch of _theca->cacheFullFlags closer to the test isn't a proper fix. Some kind of synchronization needs to be used to guarantee consistent results.

Reply:
Moved the setting of _cacheFullFlags and its comparison against
_theca->cacheFullFlags inside _runtimeFlagsProtectMutex.